synapse/storage/types.py

```diff
@@ -94,3 +94,23 @@ def __exit__(
         traceback: Optional[TracebackType],
     ) -> Optional[bool]:
         ...
+
+
+class DBAPI2Module(Protocol):
+    """The module-level attributes that we use from PEP 249.
+
+    This is NOT a comprehensive stub for the entire DBAPI2."""
+
+    __name__: str
+
+    Warning: Type[Exception]
+    Error: Type[Exception]
+    DatabaseError: Type[Exception]
+    OperationalError: Type[Exception]
+    IntegrityError: Type[Exception]
```

Comment:
Some comments on which exception types correspond to serialization failures, deadlocks, foreign key violations, etc would be greatly appreciated (motivated by https://github.com/matrix-org/synapse/blob/release-v1.59/synapse/storage/databases/main/room.py#L1080-L1096)

Also might be worth nothing that `OperationalError`s and `IntegrityError`s are all `DatabaseError`s, at least empirically.

Reply:
> Also might be worth nothing that OperationalErrors and IntegrityErrors are all DatabaseErrors, at least empirically.

Yeah, and IIRC this is mandated by DPAPI2. My first attempt was something like

```python
    DatabaseError: Type[Exception]
    OperationalError: Type[DatabaseError]
```

but mypy wasn't happy with that (can dig up the errors if you'd like). I also tried

```python
    class DatabaseError(Exception): ...
    class OperationalError(DatabaseError): ...
```

But that ran afoul of https://github.com/python/mypy/issues/6393.

Reply:
I've had a go at writing out some more comments. (Largely cribbed from Psycopg2's documentation really...)